app/estimate_square_size.go

```diff
@@ -0,0 +1,246 @@
+package app
+
+import (
+	"bytes"
+	"math"
+	"sort"
+
+	"github.com/celestiaorg/celestia-app/pkg/appconsts"
+	"github.com/celestiaorg/celestia-app/pkg/shares"
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/tendermint/tendermint/pkg/consts"
+	core "github.com/tendermint/tendermint/proto/tendermint/types"
+	coretypes "github.com/tendermint/tendermint/types"
+)
+
+// prune removes txs until the set of txs will fit in the square of size
+// squareSize. It assumes that the currentShareCount is accurate. This function
+// is far from optimal because accurately knowing how many shares any given
+// set of transactions and its message takes up in a data square that is following the
+// non-interactive default rules requires recalculating the entire square.
+// TODO: include the padding used by each msg when counting removed shares
+func prune(txConf client.TxConfig, txs []*parsedTx, currentShareCount, squareSize int) parsedTxs {
+	maxShares := squareSize * squareSize
+	if maxShares >= currentShareCount {
+		return txs
+	}
+	goal := currentShareCount - maxShares
+
+	removedContiguousShares := 0
+	contigBytesCursor := 0
+	removedMessageShares := 0
+	removedTxs := 0
+
+	// adjustContigCursor checks if enough contiguous bytes have been removed
+	// inorder to tally total contiguous shares removed
+	adjustContigCursor := func(l int) {
+		contigBytesCursor += l + shares.DelimLen(uint64(l))
+		if contigBytesCursor >= consts.TxShareSize {
```

Comment:
hmm good point, I'm unsure. We will need to make a canonical API for this and all other estimation code that we keep in the shares package in order to avoid this.

Reply:
Left an issue #700 for this, not sure how long it will take, but I think its a great idea.